### PR TITLE
MAINT: Use the PyArray_(GET|SET)_ITEM functions where possible

### PR DESCRIPTION
**Pull request by eric-wieser**

### CODE REVIEW

File: numpy/f2py/cfuncs.py [hunk 1]
```diff
@@ -331,7 +331,7 @@
 /* New SciPy */
 #define TRYPYARRAYTEMPLATECHAR case NPY_STRING: *(char *)(PyArray_DATA(arr))=*v; break;
 #define TRYPYARRAYTEMPLATELONG case NPY_LONG: *(long *)(PyArray_DATA(arr))=*v; break;
-#define TRYPYARRAYTEMPLATEOBJECT case NPY_OBJECT: (PyArray_DESCR(arr)->f->setitem)(pyobj_from_ ## ctype ## 1(*v),PyArray_DATA(arr)); break;
+#define TRYPYARRAYTEMPLATEOBJECT case NPY_OBJECT: PyArray_SETITEM(arr,PyArray_DATA(arr),pyobj_from_ ## ctype ## 1(*v)); break;
```
Comment (eric-wieser) [inline numpy/f2py/cfuncs.py]: This macro was previously a compile error, as it forgot the last argument of `setitem`. Ditto for the complex version.

File: numpy/core/src/multiarray/iterators.c [hunk 2]
```diff
@@ -917,7 +917,7 @@ iter_ass_subscript(PyArrayIterObject *self, PyObject *ind, PyObject *val)
     if (PyBool_Check(ind)) {
         retval = 0;
         if (PyObject_IsTrue(ind)) {
-            retval = type->f->setitem(val, self->dataptr, self->ao);
+            retval = PyArray_SETITEM(self->ao, self->dataptr, val);
         }
         goto finish;
     }
```
Comment (ahaldane) [inline numpy/core/src/multiarray/iterators.c]: There's another setitem call at line 938.

**Comment by ahaldane** [comment]
If we want to be complete, I see just a few more:
```
numpy/core/src/multiarray/scalarapi.c:322:        if (typecode->f->setitem(scalar, PyArray_DATA(r), r) < 0) {
numpy/core/src/multiarray/refcount.c:256:                dtype->f->setitem(obj, optr, arr);
numpy/core/src/multiarray/ctors.c:2968:    ret = funcs->setitem(obj, PyArray_DATA(range), range);
numpy/core/src/multiarray/ctors.c:2977:    ret = funcs->setitem(obj, PyArray_BYTES(range)+PyArray_ITEMSIZE(range),
numpy/core/src/multiarray/ctors.c:3190:    if (funcs->setitem(start, PyArray_DATA(range), range) < 0) {
numpy/core/src/multiarray/ctors.c:3196:    if (funcs->setitem(next, PyArray_BYTES(range)+PyArray_ITEMSIZE(range),
```

**Comment by eric-wieser** [comment]
@ahaldane: I was tricked into thinking that in those cases, the array is itself is not in scope. Obviously not true, since it's passed as the last argument

**Comment by charris** [comment]
Thanks Eric.

**Pull request merged**